lib/IDE/Refactoring.cpp

```diff
@@ -6609,20 +6676,80 @@ class AsyncConverter : private SourceEntityWalker {
+      // *either* the error *or* the result parameter is non-nil.
+
+      auto ErrorArgs =
+          TopHandler.extractResultArgs(CE, /*ReturnErrorArgsIfAmbiguous=*/true);
+      assert(ErrorArgs.isError() && "If the call was ambiguous, we should be "
+                                    "able to interpret it as an error call");
+      assert(ErrorArgs.args().size() == 1 &&
+             "There should only be a single error parameter");
+      auto SuccessArgs = TopHandler.extractResultArgs(
+          CE, /*ReturnErrorArgsIfAmbiguous=*/false);
+      assert(!SuccessArgs.isError() &&
+             "If the call was ambiguous, we should be able to interpret it as "
+             "a success call");
+
+      Identifier ErrorName = createUniqueName("error");
+      Identifier ContinuationName = Scopes.back().ContinuationName;
+
+      // if let error = theError {
+      OS << tok::kw_if << ' ' << tok::kw_let << ' ' << ErrorName << ' '
+         << tok::equal << ' ';
+      convertNode(ErrorArgs.args().back(), /*StartOverride=*/{},
+                  /*ConvertCalls=*/false);
+      OS << ' ' << tok::l_brace << '\n';
+      // assert(successParam == nil, "Expected nil success params for non-nil
+      //                              error")
+      for (auto SuccessArg : SuccessArgs.args()) {
+        OS << "assert" << tok::l_paren;
+        convertNode(SuccessArg, /*StartOverride=*/{}, /*ConvertCalls=*/false);
+        OS << " == " << tok::kw_nil << tok::comma
+           << " \"Expected nil success params for non-nil error\""
+           << tok::r_paren << '\n';
+      }
+      // continuation.resume(throwing: error)
+      OS << ContinuationName << tok::period << "resume" << tok::l_paren
+         << "throwing" << tok::colon << ' ' << ErrorName << tok::r_paren;
+      // } else {
+      //   <perform refactoring for success case>
+      // }
+      OS << '\n' << tok::r_brace << ' ' << tok::kw_else << ' ' << tok::l_brace
+         << '\n';
+      addHandlerCallToContinuationImpl(CE,
+                                       /*ReturnErrorArgsIfAmbiguous=*/false);
+      OS << '\n' << tok::r_brace;
+    } else {
+      addHandlerCallToContinuationImpl(CE, /*ReturnErrorArgsIfAmbiguous=*/true);
+    }
+  }
+
   /// Assuming that \p CE is a call to \c TopHandler's completion handler and
   /// that the current scope is wrapped in a continuation, replace it with a
   /// call to the continuation.
-  void addHandlerCallToContinuation(const CallExpr *CE) {
+  void addHandlerCallToContinuationImpl(const CallExpr *CE,
+                                        bool ReturnErrorArgsIfAmbiguous) {
     assert(TopHandler.getAsHandlerCall(const_cast<CallExpr *>(CE)) == CE &&
            "addHandlerCallToContinuation must be used with a call to the "
            "TopHandler's completion handler");
     assert(Scopes.back().isWrappedInContination());
 
-    ArrayRef<Expr *> Args;
+    std::vector<Expr *> Args;
```

Comment:
Out of curiosity, why did this need to become a vector?

Reply:
That’s because actually before this might result in unsafe memory access. 🙈 

In the params case, `Args` references into `Exprs`, which goes out of scope before `Args`. Making it a `std::vector` makes sure that all `Expr` pointers stay alive. I already cherry-picked this fix to the 5.5 PR #38256.

Reply:
Ah, I missed the fact that `HandlerResult` can own the elements in the single element case, good catch!